### PR TITLE
Refactor action to support upload at root, glob matching

### PR DESCRIPTION
**Pull request by bharathkkb**
This PR

- Adds a parent option which allows users to omit parent from destination path of the uploaded object #11 
- Adds glob matching support #20
- Improves documentation for various flags #23
- Improves testing

### CODE REVIEW

File: src/upload-helper.ts [hunk 1]
```diff
@@ -56,58 +61,64 @@ export class UploadHelper {
...
-          filePath,
-          gzip,
-          destination,
-          predefinedAcl,
-        );
-        return uploadResp;
-      }),
-    );
-    return resp;
+    // const filesList = await getFiles(directoryPath,parent);
```
Comment (averikitsch) [inline src/upload-helper.ts:104]: nit: comment

File: src/upload-helper.ts [hunk 2]
```diff
@@ -56,58 +61,64 @@ export class UploadHelper {
...
+    const filesList = await globby([path.posix.join(directoryPath, glob)]);
```
Comment (averikitsch) [inline src/upload-helper.ts:105]: nit: can you add a comment here. I assume if glob = '' it returns all files?

File: tests/upload.int.test.ts [hunk 3]
```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2020 Google LLC
```
Comment (averikitsch) [inline tests/upload.int.test.ts:2]: nit: 2021

File: tests/upload.int.test.ts [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+  FILES_IN_DIR,
+  EXAMPLE_PREFIX,
+  FILES_IN_DIR_WITHOUT_PARENT_DIR,
+  TXT_FILES_IN_DIR,
+  TXT_FILES_IN_DIR_WITHOUT_PARENT_DIR,
+  TXT_FILES_IN_TOP_DIR,
+} from './constants.test';
+
+const storage = new Storage({
+  projectId: process.env.UPLOAD_CLOUD_STORAGE_TEST_PROJECT,
```
Comment (averikitsch) [inline tests/upload.int.test.ts:39]: I don't see this env var passed into the test via the workflow?

**Review by averikitsch — APPROVED** [review]

**Comment by williammartin** [comment]
For what it's worth, I just discovered I needed this so I'd be pretty keen to try it out and also review if that would help you get it in. Cheers 👏

**Comment by bharathkkb** [comment]
@williammartin a review would be great, I'll try to fix conflicts and get this in soon

**Comment by williammartin** [comment]
Sure I'll have a look after you fix conflicts. FWIW I forked this and used the `parent` field in my workflow and it worked like I hoped.

File: package.json [hunk 5]
```diff
@@ -6,8 +6,9 @@
   "scripts": {
     "build": "ncc build src/main.ts",
     "lint": "eslint . --ext .ts,.tsx",
+    "fixlint": "eslint . --ext .ts,.tsx --fix",
     "format": "prettier --write **/*.ts",
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply package.json]: TIL, done

File: src/main.ts [hunk 6]
```diff
@@ -27,6 +27,10 @@ async function run(): Promise<void> {
     const predefinedAclInput = core.getInput('predefinedAcl', {
       required: false,
     });
+    const parent =
+      core.getInput('parent', { required: false }) === 'false' ? false : true;
```
[the remark this answers is not in the document]
Comment (bharathkkb) [reply src/main.ts]: done

File: src/upload-helper.ts (hunk 2 above)
Comment (bharathkkb) [reply src/upload-helper.ts:120]: added a comment, `glob = ''` evaluates `path.posix.join(directoryPath, glob)` to just `directoryPath/` which then globby uses as `directoryPath/**/*`. This is default behavior where previously we were recursively walking the dir.

File: tests/upload.int.test.ts (hunk 3 above)
Comment (bharathkkb) [reply tests/upload.int.test.ts]: done

File: tests/upload.int.test.ts (hunk 4 above)
Comment (bharathkkb) [reply tests/upload.int.test.ts:39]: oops added it. Some tests were silently being skipped.

**Pull request merged**